net.c

```diff
@@ -7,143 +7,237 @@
-        continue;
-      }
-      r = setsockopt(fd, SOL_SOCKET, SO_KEEPALIVE, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting SO_KEEPALIVE on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, SOL_SOCKET, SO_LINGER, &linger, sizeof linger);
-      if (r == -1) {
-        twarn("setting SO_LINGER on fd %d", fd);
-        close(fd);
-        continue;
-      }
-      r = setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &flags, sizeof flags);
-      if (r == -1) {
-        twarn("setting TCP_NODELAY on fd %d", fd);
+        r = bind(fd, ai->ai_addr, ai->ai_addrlen);
+        if (r == -1) {
+            twarn("bind()");
+            close(fd);
+            continue;
+        }
+        if (verbose) {
+            char hbuf[NI_MAXHOST], pbuf[NI_MAXSERV], *h = host, *p = port;
+            struct sockaddr_in addr;
+            socklen_t addrlen;
+
+            addrlen = sizeof(addr);
+            r = getsockname(fd, (struct sockaddr *) &addr, &addrlen);
+            if (!r) {
+                r = getnameinfo(ai->ai_addr, ai->ai_addrlen,
+                                hbuf, sizeof hbuf,
+                                pbuf, sizeof pbuf,
+                                NI_NUMERICHOST|NI_NUMERICSERV);
+                if (!r) {
+                    h = hbuf;
+                    p = pbuf;
+                }
+            }
+            if (ai->ai_family == AF_INET6) {
+                printf("bind %d [%s]:%s\n", fd, h, p);
+            } else {
+                printf("bind %d %s:%s\n", fd, h, p);
+            }
+        }
+
+        r = listen(fd, 1024);
+        if (r == -1) {
+            twarn("listen()");
+            close(fd);
+            continue;
+        }
+
+        break;
+    }
+
+    freeaddrinfo(airoot);
+
+    if(ai == NULL)
+        fd = -1;
+
+    return fd;
+}
+
+static int
+make_unix_socket(char *path)
+{
+    int fd = -1, r;
+    struct stat st;
+    struct sockaddr_un addr;
+    const size_t maxlen = sizeof(addr.sun_path) - 1; // Reserve the last position for '\0'
+
+    r = stat(path, &st);
+    if (r == 0) {
+        if (S_ISSOCK(st.st_mode)) {
+            warnx("removing existing local socket to replace it");
+            r = unlink(path);
+            if (r == -1) {
+                twarn("unlink");
+                return -1;
+            }
+        } else {
+            twarnx("another file already exists in the given path");
+            return -1;
+        }
+    }
+
+    fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        twarn("socket()");
+        return -1;
+    }
+
+    r = set_nonblocking(fd);
+    if (r == -1) {
         close(fd);
-        continue;
-      }
+        return -1;
+    }
 
-      r = bind(fd, ai->ai_addr, ai->ai_addrlen);
-      if (r == -1) {
+    memset(&addr, 0, sizeof(struct sockaddr_un));
+    addr.sun_family = AF_UNIX;
+    if (strlen(path) > maxlen) {
+        warnx("truncating local socket path to: %.*s", (int)maxlen, path);
```

Comment:
I propose that we `return -1` in case of truncated socket path. We should fail clearly ASAP of socket path is not correct. The test looked fishy, so I examined behaviour here, we should not allow this kind of error because when path was truncated then user definitely has an issue. We can keep the `strncpy` as safety measure though.

And test can check the returned results to check that long socket path or path to wrong directory won't be tolerated by the program.

Reply:
Totally agree.  Even with a warning, truncating the name may just lead to other unrelated and puzzling failures.

I didn't think that thoroughly before.